Libraries/Opc.Ua.Client/Session/SessionAsync.cs

```diff
@@ -1051,6 +1051,294 @@ List<ServiceResult> errors
+            // optimize later, when everything works fine.
+            for (int i = 0; i < nodesToBrowse.Count; i++)
+            {
+                result.Add(new ReferenceDescriptionCollection());
+                errors.Add(new ServiceResult(StatusCodes.Good));
+            }
+            try
+            {
+                // in the first pass, we browse all nodes from the input.
+                // Some nodes may need to be browsed again, these are then fed into the next pass.
+                List<NodeId> nodesToBrowseForPass = new List<NodeId>();
+                nodesToBrowseForPass.AddRange(nodesToBrowse);
+
+                List<ReferenceDescriptionCollection> resultForPass = new List<ReferenceDescriptionCollection>();
+                resultForPass.AddRange(result);
+
+                List<ServiceResult> errorsForPass = new List<ServiceResult>();
+                errorsForPass.AddRange(errors);
+
+                int passCount = 0;
+
+                do
+                {
+                    int badNoCPErrorsPerPass = 0;
+                    int badCPInvalidErrorsPerPass = 0;
+                    int otherErrorsPerPass = 0;
+                    uint maxNodesPerBrowse = OperationLimits.MaxNodesPerBrowse;
+
+                    if (executeDefensively && ServerMaxContinuationPointsPerBrowse > 0)
+                    {
+                        maxNodesPerBrowse = ServerMaxContinuationPointsPerBrowse < maxNodesPerBrowse ? ServerMaxContinuationPointsPerBrowse : maxNodesPerBrowse;
+                    }
+
+                    // split input into batches
+                    int batchCount = 0;
+
+                    List<NodeId> nodesToBrowseForNextPass = new List<NodeId>();
+                    List<ReferenceDescriptionCollection> referenceDescriptionsForNextPass = new List<ReferenceDescriptionCollection>();
+                    List<ServiceResult> errorsForNextPass = new List<ServiceResult>();
+
+                    // loop over the batches
+                    foreach (var nodesToBrowseBatch in ((List<NodeId>)nodesToBrowseForPass).Batch<NodeId, List<NodeId>>(maxNodesPerBrowse))
+                    {
+                        int batchOffset = batchCount * (int)maxNodesPerBrowse;
+
+                        (
+                            List<ReferenceDescriptionCollection> resultForBatch,
+                            IList<ServiceResult> errorsForBatch
+                        )
+                        =
+                        await BrowseWithBrowseNextAsync(
+                                view,
+                                nodesToBrowseBatch,
+                                maxResultsToReturn,
+                                browseDirection,
+                                referenceTypeId,
+                                includeSubtypes,
+                                nodeClassMask
+                            ).ConfigureAwait(false);
+
+                        for (int ii = 0; ii < nodesToBrowseBatch.Count; ii++)
```

Comment:
can some common code with the sync version be extracted to private methods?

Reply:
I removed the non-async method's bodies and now call the async methods in there. So there is no duplicate coding of that kind anymore.